[WFLY-17144] Add Micrometer support to WildFly

jasondlee opened the pull request:
https://issues.redhat.com/browse/WFLY-17144

Move configs from Preview to Standard
Migrate modules
Add OTLP support
Remove prometheus support and dependencies
Add 'step' config parameter
Clean up Micrometer API usage
Module clean up and reorg

yersan commented:
relates to https://github.com/wildfly/wildfly-core/pull/5367

bstansberry commented on preview/feature-pack/wildfly-feature-pack-build.xml:
```diff
@@ -85,7 +85,6 @@
                 <extension>org.wildfly.extension.io</extension>
                 <extension>org.wildfly.extension.messaging-activemq</extension>
                 <extension>org.wildfly.extension.metrics</extension>
-                <extension>org.wildfly.extension.micrometer</extension>
```
Please add this back.

jasondlee replied:
Brian ended up addressing this in a PR against this PR. :)

jasondlee replied:
Brian ended up addressing this in a PR against this PR. :)

yersan commented on ee-feature-pack/common/src/main/resources/modules/system/layers/base/io/opentelemetry/otlp/main/module.xml:
```diff
@@ -21,13 +19,22 @@
   -->
-
-<module xmlns="urn:jboss:module:1.9" name="org.latencyutils">
+<module xmlns="urn:jboss:module:1.9" name="io.opentelemetry.otlp">
```
just out of curiosity, why are we adding micrometer related dependencies and modules to the ee-standard if the micrometer extension is only available via galleon-pack?

jasondlee replied:
The OpenTelemetry modules are already a part of the ee-standard FP. This Micrometer change requires some additional OTLP support, so, rather that bifurcating that configuration, I've opted to keep all of the OTLP-related module definitions together.

jasondlee replied:
@bstansberry Does this need to be done for WF 28 code freeze?

bstansberry replied:
@jasondlee Yes. Not for the beta though.

jasondlee replied:
@bstansberry, So not this week, but very soon after, then? If so, I'll file a follow up JIRA upon confirmation. :)

yersan commented on ee-feature-pack/galleon-common/src/main/resources/feature_groups/domain-host-excludes.xml:
```diff
@@ -5,90 +5,95 @@
     <feature spec="domain.host-exclude">
         <param name="host-exclude" value="WildFly10.0"/>
-        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.clustering.ejb&quot;,&quot;org.wildfly.extension.clustering.web&quot;,&quot;org.wildfly.extension.core-management&quot;,&quot;org.wildfly.extension.datasources-agroal&quot;,&quot;org.wildfly.extension.discovery&quot;,&quot;org.wildfly.extension.ee-security&quot;,&quot;org.wildfly.extension.elytron&quot;,&quot;org.wildfl ...
+        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.clustering.ejb&quot;,&quot;org.wildfly.extension.clustering.web&quot;,&quot;org.wildfly.extension.core-management&quot;,&quot;org.wildfly.extension.datasources-agroal&quot;,&quot;org.wildfly.extension.discovery&quot;,&quot;org.wildfly.extension.ee-security&quot;,&quot;org.wildfly.extension.elytron&quot;,&quot;org.wildfl ...
```
It is harmless since the micrometer extension is only available via galleon-pack but these exclusions here are unexpected. At this moment, from the point of view of an ee-standard build, the micrometer extension does not exist.

jasondlee replied:
Updated.

yersan commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/HostExcludesTestCase.java:
```diff
@@ -208,7 +210,11 @@ private enum ExtensionConf {
         // Don't add here extensions supplied only by the wildfly-preview-feature-pack because we are not tracking different releases
         // of wildfly preview. In such a case, add them to previewExtensions set defined below.
-        CURRENT(MAJOR, WILDFLY_27_0, null, getCurrentRemovedExtensions());
+        CURRENT(MAJOR, WILDFLY_27_0, Arrays.asList(
+                "org.wildfly.extension.micrometer"
+        ),
+                getCurrentRemovedExtensions()
+        );
```
Notice the comment about this enum:

```
If an extension is added to this enum, also check if it is supplied only by wildfly-galleon-pack. If so, add it also
to the internal mpExtensions Set defined on this class.
```

You also add the "org.wildfly.extension.micrometer" to the `mpExtensions` HashSet.

yersan commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/observability/micrometer/MicrometerSetupTask.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Override
+    public void setup(final ManagementClient managementClient, final String containerId) throws Exception {
+        execute(managementClient, Operations.createAddOperation(extensionAddress), true);
+        execute(managementClient, Operations.createAddOperation(subsystemAddress), true);
```
Should not the metrics extension/subsystem be removed to perform this test? The documentation describes that metrics should be removed to use micrometer, so I'm pointing out here in case this test could fail because both extensions are added at the same time.

jasondlee replied:
The docs are actually a bit out of date. These two extensions can be run together, though I image there would be some performance impact. It wouldn't hurt, I guess, to remove it, as that is (or will be) the suggested configuration.

rhusar commented:
I don't think the core upgrade is needed any more; removing stale flag.

yersan reviewed (changes requested):
Added minor points and doubts I found in the first round. I would also expect a change in the documentation since it describes this subsystem as only valid for preview: 

https://docs.wildfly.org/27/Admin_Guide.html#Micrometer_Metrics

Since it seems this extension cannot be added with metrics, it would be nice to describe in the Micrometer the management operations that remove the metrics subsystem

yersan commented on preview/galleon-content/src/main/resources/feature_groups/standalone-ha.xml:
```diff
@@ -21,7 +21,6 @@
     <feature-group name="management-interfaces"/>
     <feature-group name="access-control"/>
     <feature-group name="health"/>
-    <feature-group name="micrometer" />
```
This should be replaced with "metrics", otherwise the non-microprofile standalone configurations will lack something to get metrics in the preview distribution.

yersan reviewed (changes requested):
Found two details on a second round

jasondlee commented on preview/galleon-content/src/main/resources/feature_groups/standalone-profile.xml:
```diff
@@ -5,6 +5,5 @@
 
     <feature-group name="basic-profile"/>
     <feature-group name="health"/>
-    <feature-group name="micrometer" />
```
IIRC, there is already a "metrics" layer, which is where WildFly Metrics lives, and that's not going away at this point.

jasondlee commented on boms/standard-expansion/pom.xml:
```diff
@@ -678,6 +677,107 @@
+            <dependency>
+                <groupId>io.opentelemetry.proto</groupId>
+                <artifactId>opentelemetry-proto</artifactId>
+                <version>0.19.0-alpha</version>
```
Fixed.

rhusar commented on testsuite/integration/basic/src/test/java/org/wildfly/test/integration/observability/micrometer/MetricResource.java:
```diff
@@ -19,25 +20,25 @@ public class MetricResource {
-        }
-        counter.increment();
-        sample.stop(timer);
+        Timer timer = meterRegistry.timer("demo_timer", Tags.of("ts", ""+System.currentTimeMillis()));
```
Please run formatter on this class.

jasondlee replied:
Fixed

rhusar commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemDefinition.java:
```diff
@@ -40,54 +42,58 @@ class MicrometerSubsystemDefinition extends PersistentResourceDefinition {
+
+    public static final SimpleAttributeDefinition STEP = SimpleAttributeDefinitionBuilder
+            .create(MicrometerConfigurationConstants.STEP, ModelType.LONG, true)
+            .setDefaultValue(ModelNode.fromString("60")) // 1 minute
```
Unnecessary parsing here, just use `.setDefaultValue(new ModelNode(60))` istead

rhusar replied:
Or even better `new ModelNode(TimeUnit.MINUTES.toSeconds(1)`

rhusar commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerRegistryService.java:
```diff
@@ -34,32 +34,36 @@
-     * @param context the management operation context to use to install the service. Cannot be {@code null}
+     *
+     * @param context  the management operation context to use to install the service. Cannot be {@code null}
+     * @param config
```
BTW missing javadoc description here

rhusar commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemDefinition.java:
```diff
@@ -40,54 +42,58 @@ class MicrometerSubsystemDefinition extends PersistentResourceDefinition {
+            .setDefaultValue(ModelNode.fromString("60")) // 1 minute
+            .setAllowExpression(true)
+            .setRestartAllServices()
+            .build();
```
Is this missing `.setMeasurementUnit(MeasurementUnit.SECONDS)`?

rhusar commented on observability/micrometer/src/main/resources/org/wildfly/extension/micrometer/LocalDescriptions.properties:
```diff
@@ -5,3 +5,5 @@ micrometer.remove=Operation Removes subsystem
 micrometer.prefix=Prefix prepended to the name of the WildFly metrics exposed by the HTTP endpoints.
 micrometer.security-enabled=True if authentication is required to access the HTTP endpoint on the HTTP management interface.
+micrometer.endpoint=The address to where metrics will be published. The OTLP default is http://localhost:4318/v1/metrics, though WildFly provides no default, so this must be configured manually.
+micrometer.step=The step size (reporting frequency) to use in seconds. The default is 60 seconds.
```
Remove "The default is 60 seconds." from the description. The default value is in the model, e.g.:

```
            "step" => {
                "type" => LONG,
                "description" => "The step size (reporting frequency) to use in seconds. The default is 60 seconds.",
                "expressions-allowed" => true,
                "required" => false,
                "nillable" => true,
                "default" => 60,
                "access-type" => "read-write",
                "storage" => "configuration",
                "restart-required" => "all-services"
            }
```

jasondlee replied:
Changed.

rhusar commented on observability/micrometer/src/main/resources/org/wildfly/extension/micrometer/LocalDescriptions.properties:
```diff
@@ -5,3 +5,5 @@ micrometer.remove=Operation Removes subsystem
 micrometer.exposed-subsystems=The names of the WildFly subsystems that exposes their metrics (or '*' to expose any subsystem metrics).
 micrometer.prefix=Prefix prepended to the name of the WildFly metrics exposed by the HTTP endpoints.
 micrometer.security-enabled=True if authentication is required to access the HTTP endpoint on the HTTP management interface.
+micrometer.endpoint=The address to where metrics will be published. The OTLP default is http://localhost:4318/v1/metrics, though WildFly provides no default, so this must be configured manually.
```
This is wrong. When I boot the server without any manual configuration I am getting failures to connect.

```
[standalone@embedded /] 17:46:00,011 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:177)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:507)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:602)
	at java.base/sun.net.www.http.HttpClient.<init>(HttpClient.java:275)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:374)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:395)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1253)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:1015)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1367)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1342)
	at io.micrometer@1.9.3//io.micrometer.core.ipc.http.HttpUrlConnectionSender.send(HttpUrlConnectionSender.java:98)
	at io.micrometer@1.9.3//io.micrometer.core.ipc.http.HttpSender$Request$Builder.send(HttpSender.java:305)
	at io.micrometer@1.9.3//io.micrometer.registry.otlp.OtlpMeterRegistry.publish(OtlpMeterRegistry.java:109)
	at io.micrometer@1.9.3//io.micrometer.core.instrument.push.PushMeterRegistry.publishSafely(PushMeterRegistry.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

jasondlee replied:
You're right. The code has changed since I wrote that. I'll update it.

rhusar commented on observability/micrometer/src/main/resources/schema/wildfly-micrometer_1_0.xsd:
```diff
@@ -26,17 +26,26 @@
                 <xs:annotation>
                     <xs:documentation>
-                        The names of the subsystems (separated by spaces) that exposes their metrics in the vendor scope (or '*' to expose any subsystem metrics).
+                        The step size (reporting frequency) to use in seconds. The default is 60 seconds.
```
Also, remove the "The default is 60 seconds." from here. And use XSD to define it.

rhusar commented on observability/micrometer/src/main/resources/schema/wildfly-micrometer_1_0.xsd:
```diff
@@ -26,17 +26,26 @@
+                    </xs:documentation>
+                </xs:annotation>
+            </xs:attribute>
+            <xs:attribute name="step" type="xs:string">
```
Fix:
```
            <xs:attribute name="step" type="xs:string" default="60">
```

rhusar commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemDefinition.java:
```diff
@@ -40,54 +44,59 @@ class MicrometerSubsystemDefinition extends PersistentResourceDefinition {
     private static final String MICROMETER_MODULE = "org.wildfly.extension.micrometer";
     private static final String MICROMETER_API_MODULE = "org.wildfly.extension.micrometer-api";
```
The Micrometer API module name is definitely wrong. This is not a WF extension in the sense we use this word. Also applies to java package name but I can't comment on that line since it's not being changed.

How about "org.wildfly.micrometer.deployment" ?

rhusar replied:
To save anyone time who might be looking at this: this module contains single CDI extension class (MicrometerCdiExtension) which registers the MeterRegistry type.

jasondlee replied:
Changed.

rhusar commented:
There's another problem , the OTLP publishers are never closed on reload! For a simple test, shutdown the registry process, and if you just reload the server 10 times you end up with 10 metrics publishers trying to connect and publish.

```
[standalone@embedded /] 18:18:59,934 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-2) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:18:59,934 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:19:00,003 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-8) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:19:00,003 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-6) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:19:00,003 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-7) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:19:00,007 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-4) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:19:00,008 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-3) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:19:00,008 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-5) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
18:19:00,012 WARNING [io.micrometer.registry.otlp.OtlpMeterRegistry] (otlp-metrics-publisher-9) Failed to publish metrics to OTLP receiver: java.net.ConnectException: Connection refused (Connection refused)
```

jasondlee commented:
> There's another problem , the OTLP publishers are never closed on reload! For a simple test, shutdown the registry process, and if you just reload the server 10 times you end up with 10 metrics publishers trying to connect and publish every minute.

Nice catch. Fixed locally.

rhusar commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Micrometer.adoc:
```diff
@@ -91,34 +45,23 @@ The HTTP endpoint exposes the following metrics:
-of the `Servlet` are added to the tags of the metric.
+For example Undertow exposes a metric attribute `request-count` for every Servlet in an application deployment. This attribute will be exposed with the name `undertow_request_count`. Other information such as the name of the `Servlet` are added to the tags of the metric.
+
+The https://github.com/wildfly/quickstart/tree/main/helloworld[helloworld quickstart] demonstrates the use of CDI and Servlet in Wildfly. A corresponding metric will be exposed for it with the name and tags:
```
s/Wildfly/WildFly

jasondlee commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/HostExcludesTestCase.java:
```diff
@@ -208,24 +210,29 @@ private enum ExtensionConf {
-                        "org.wildfly.extension.picketlink",
-                        "org.jboss.as.security"
+                        // This is moving from Preview to Standard in WF 28, so it's both being removed and added.
+                    "org.wildfly.extension.micrometer"
```
@yersan @bstansberry, this feels hinky, but it works. If there's a more correct way, please chime in.

bstansberry commented on galleon-pack/common/src/main/resources/license/wildfly-feature-pack-common-licenses.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<licenseSummary>
+  <dependencies>
+    <!-- TODO add license data for the dependencies in this module's pom -->
```
Needs to be done. :)

bstansberry commented on boms/standard-ee/pom.xml:
```diff
@@ -738,6 +738,63 @@
+                        <artifactId>*</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
```
Why are these in this bom instead of standard-expansion?

jasondlee replied:
Probably an early change before I understood the differences and before your changes to make deps defined here more shareable? Regardless, cleaned it up.

bstansberry commented on testsuite/integration/basic/pom.xml:
```diff
@@ -805,6 +805,11 @@
             <artifactId>jakarta.servlet.jsp-api</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-core</artifactId>
+            <scope>test</scope>
+        </dependency>
```
@yersan @jasondlee @parsharma This is not a testbom dependency. It's a normal production code dependency already managed via the production code boms and part of the io.micrometer JBoss Modules module.

bstansberry reviewed (commented):
You need to make the same changes to preview/feature-pack/pom.xml that you made to galleon-pack/galleon-feature-pack/pom.xml.  You've created a new maven module for shared Galleon content and you need to include that content in the build of the wildfly-preview feature pack.

I'll comment separately about HostExcludesTestCase.

jasondlee commented on galleon-pack/galleon-feature-pack/wildfly-feature-pack-build.xml:
```diff
@@ -70,6 +70,7 @@
     <generate-feature-specs>
         <extensions>
             <standalone>
+                <extension>org.wildfly.extension.micrometer</extension>
```
Done

jasondlee commented:
/retest

bstansberry commented on galleon-pack/common/src/main/resources/license/wildfly-feature-pack-common-licenses.xml:
```diff
@@ -0,0 +1,60 @@
+        </dependency>
+        <dependency>
+            <groupId>com.google.protobuf</groupId>
+            <artifactId>protobuf-java</artifactId>
```
This entry needs to be first.

bstansberry commented on observability/micrometer-api/pom.xml:
```diff
@@ -63,12 +63,22 @@
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
```
@jasondlee Please remove the exclusions.

bstansberry commented on observability/micrometer-api/pom.xml:
```diff
@@ -63,12 +63,22 @@
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
```
@jasondlee Please remove the exclusions.

jasondlee replied:
Removed

bstansberry commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerConfigurationConstants.java:
```diff
@@ -0,0 +1,9 @@
+package org.wildfly.extension.micrometer;
```
Needs the copyright/license header.

jasondlee replied:
Added

bstansberry commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerDeploymentService.java:
```diff
@@ -40,22 +40,28 @@
     private final boolean exposeAnySubsystem;
     private final List<String> exposedSubsystems;
 
+    private MetricRegistration registration;
```
This should be volatile.

bstansberry commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerRegistryService.java:
```diff
@@ -34,32 +34,37 @@
 class MicrometerRegistryService implements Service {
     private final Consumer<WildFlyRegistry> registriesConsumer;
+    private final WildFlyMicrometerConfig config;
+    private WildFlyRegistry registry;
```
volatile

bstansberry commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/WildFlyMicrometerConfig.java:
```diff
@@ -0,0 +1,55 @@
+
+import io.micrometer.registry.otlp.OtlpConfig;
+
+public class WildFlyMicrometerConfig implements OtlpConfig {
```
final

bstansberry commented on observability/micrometer/src/main/resources/org/wildfly/extension/micrometer/LocalDescriptions.properties:
```diff
@@ -5,3 +5,5 @@ micrometer.remove=Operation Removes subsystem
 micrometer.exposed-subsystems=The names of the WildFly subsystems that exposes their metrics (or '*' to expose any subsystem metrics).
 micrometer.prefix=Prefix prepended to the name of the WildFly metrics exposed by the HTTP endpoints.
 micrometer.security-enabled=True if authentication is required to access the HTTP endpoint on the HTTP management interface.
+micrometer.endpoint=The address to where metrics will be published.
```
s/address/URL ?

jasondlee replied:
Changed

bstansberry commented on observability/micrometer/src/main/resources/schema/wildfly-micrometer_1_0.xsd:
```diff
@@ -1,5 +1,5 @@
 <!--
-  ~ Copyright 2022 Red Hat, Inc.
+  ~ Copyright 2022-2023 Red Hat, Inc.
```
No need to change, but don't bother updating these dates. We've been told by those who know that this isn't necessary.

bstansberry commented on observability/micrometer/src/main/resources/schema/wildfly-micrometer_1_0.xsd:
```diff
@@ -26,17 +26,25 @@
+                    </xs:documentation>
+                </xs:annotation>
+            </xs:attribute>
+            <xs:attribute name="step" type="xs:string" default="60">
```
This is xs:int

jasondlee replied:
Good catch. Fixed.

bstansberry commented on testsuite/integration/basic/src/test/resources/arquillian.xml:
```diff
@@ -4,6 +4,10 @@
 
+    <engine>
+        <property name="deploymentExportPath">target/deployments</property>
+    </engine>
```
Is this cruft?

rhusar commented on observability/micrometer/src/main/resources/org/wildfly/extension/micrometer/LocalDescriptions.properties:
```diff
@@ -5,3 +5,5 @@ micrometer.remove=Operation Removes subsystem
 micrometer.exposed-subsystems=The names of the WildFly subsystems that exposes their metrics (or '*' to expose any subsystem metrics).
 micrometer.prefix=Prefix prepended to the name of the WildFly metrics exposed by the HTTP endpoints.
 micrometer.security-enabled=True if authentication is required to access the HTTP endpoint on the HTTP management interface.
```
These two descriptions `prefix` and `security-enabled` aren't used any more.

rhusar commented on observability/micrometer/src/main/resources/schema/wildfly-micrometer_1_0.xsd:
```diff
@@ -26,17 +26,25 @@
                 <xs:annotation>
                     <xs:documentation>
-                        True if authentication is required to access the HTTP endpoint on the HTTP management interface.
+                        The names of the subsystems (separated by spaces) that exposes their metrics in the vendor scope
+                        (or '*' to expose any subsystem metrics).
```
Trivial, but in this situation, we should just keep these descriptions in sync with the `localdescriptions.properties`.

jasondlee committed:
[WFLY-17144] Add Micrometer support to WildFly

Move configs from Preview to Standard
Migrate modules
Add OTLP support
Remove prometheus support and dependencies
Add 'step' config parameter
Clean up Micrometer API usage
Module clean up and reorg
Include the galleon-pack/common content in the wildfly-preview feature pack
Get HostExcludesTestCase passing
standard-expansion bom cleanups -- remove duplicate dep declarations, use correct sorting

fabiobrz commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Micrometer.adoc:
```diff
@@ -9,87 +9,44 @@ ifdef::env-github[]
-If the authentication fails, the server will reply with a `401 NOT AUTHORIZED` response.
+* `endpoint` - the URL of the metrics collector endpoint (default: http://localhost:4318/v1/metrics)
+* `exposed-subsystems` - a list of strings corresponding to the names of subsystems that exposes their metrics in the
+HTTP metrics endpoints. By default, it is not defined (there will be no metrics exposed by subsystem. The special wildcard "{asterisk}" can be used to expose metrics from _all_ subsystems. The standalone configuration sets this attribute to "{asterisk}".
```
Minor typo: missing the closing `)` here after `exposed by subsystem`.

jasondlee replied:
Fixed

fabiobrz reviewed (commented):
Hi @jasondlee - I've had a look to the community docs, that LGTM, I just spotted a minor typo and left a question about the subsystem configuration CLI statements.
Feel free to let me know what you think.

rhusar commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Micrometer.adoc:
```diff
@@ -9,87 +9,44 @@ ifdef::env-github[]
 [source,options="nowrap"]
 ----
 [standalone@localhost:9990 /] /extension=org.wildfly.extension.micrometer:add
+[standalone@localhost:9990 /] /subsystem=micrometer:write-attribute(name=endpoint,value=http://localhost:4318/v1/metrics)
```
Nice find - yeah this won't work. Either needs to be moved below or the attribute be specidied in the :add operation.

jasondlee replied:
You're right, @fabiobrz. I think I have that fixed locally and forgot to push, but, yes, it's wrong. Incoming... :)

jasondlee committed:
[WFLY-17144]

fabiobrz reviewed (approved):
Thanks @jasondlee - I've reviewed your changes to the community docs and those LGTM.

rhusar commented on observability/micrometer/src/main/resources/schema/wildfly-micrometer_1_0.xsd:
```diff
@@ -26,17 +26,24 @@
                 </xs:annotation>
             </xs:attribute>
-            <xs:attribute name="exposed-subsystems" type="xs:string">
+            <xs:attribute name="endpoint" type="xs:string" default="http://localhost:4318/v1/metrics">
```
The defaultvalue is now inconsistent with `org.wildfly.extension.micrometer.MicrometerSubsystemDefinition#ENDPOINT`. :-(

rhusar commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemDefinition.java:
```diff
@@ -98,9 +107,7 @@ public Collection<AttributeDefinition> getAttributes() {
-                RuntimePackageDependency.required("io.micrometer" ),
-                RuntimePackageDependency.required("io.prometheus" ),
-                RuntimePackageDependency.required("org.latencyutils" )
+                RuntimePackageDependency.required("io.micrometer" )
```
Please run the formatter on this class (we don't use extra spaced before `)`).

rhusar commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemDefinition.java:
```diff
@@ -38,56 +42,61 @@
-                    .setAllowExpression(true)
-                    .build();
+    public static final SimpleAttributeDefinition ENDPOINT = SimpleAttributeDefinitionBuilder
+            .create(MicrometerConfigurationConstants.ENDPOINT, ModelType.STRING, true)
```
Just a quick question - if this is now optional - is there a legitimate use case for the subsystem without publishing to an endpoint?

jasondlee replied:
It's not optional (see next line). @bstansberry's concern was that we were baking into the model a default value based on a... best practice set upstream that could change, so we leave off the default, but make it required, so that the admin has to provide the correct value at the time.

rhusar commented on observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemDefinition.java:
```diff
@@ -38,56 +42,61 @@
-                    .build();
+    public static final SimpleAttributeDefinition ENDPOINT = SimpleAttributeDefinitionBuilder
+            .create(MicrometerConfigurationConstants.ENDPOINT, ModelType.STRING, true)
+            .setRequired(true)
```
No need to set optional true in create(..) and then setRequired(true) :-) just use

`.create(MicrometerConfigurationConstants.ENDPOINT, ModelType.STRING)`

jasondlee replied:
Yeah, sorry. That was an oversight. I'll get that cleaned up.
